### ISSUE TITLE
Keep input makes the RevNet loss NaN

### ISSUE THREAD

**Issue by CorcovadoMing** [silvandeleemput/memcnn#29]
* MemCNN version: Source from master branch
* Python version: 3.6 (PyTorch 1.3.1)
* Operating System: Ubuntu 18.04

### Description

Hi, I easily get NaN loss while training RevNet110, and it sometimes happen while training RevNet38, the workaround is to keep the input by comment out https://github.com/silvandeleemput/memcnn/blob/master/memcnn/models/revop.py#L112

However, it makes no sense when using RevNet and still need to keep the input. Any suggestion or solution to this issue? Thanks

### What I Did

```
python -m memcnn.train revnet110 cifar10
```

Outputs

```
19-12-04 09:19:11 [trainer        ] Model parameters: 1854890
19-12-04 09:19:11 [trainer        ] Model memory allocation: 7657472
19-12-04 09:19:23 [trainer        ] iteration: [100/80000]      Time 0.083 (0.095)      Data 0.002 (0.027)      Loss nan (nan)  Prec@1 7.000 (3.390)    ActMem 105110528.000 (105310198.774)
19-12-04 09:19:23 [root           ] NaN or Inf found in input tensor.
```

**Comment by silvandeleemput** [comment]
@CorcovadoMing Hi, thank you for reporting this issue. I haven't tested MemCNN yet with PyTorch 1.3.1 so it might be a new issue. You could test with an older version of PyTorch (highest tested PyTorch version is 1.1) and see if the issue persists.

I agree that you should definitely be able to use RevNet without discarding the input.

**Comment by CorcovadoMing** [comment]
Hi @silvandeleemput,

Thanks for your quick response

However, I've just tested with PyTorch 1.1 and the issue is still there
The minimal steps to reproduce this issue is using docker

Image: `pytorch/pytorch:1.1.0-cuda10.0-cudnn7.5-runtime`
install the minimal dependencies `pip install tensorflow tensorboardX`
Then, pull the source and execute `python -m memcnn.train revnet110 cifar10` will show NaN loss immediately

**Comment by silvandeleemput** [comment]
@CorcovadoMing Thanks for checking. I'll look into this issue tomorrow.

**Comment by silvandeleemput** [comment]
@CorcovadoMing I have been working on this issue, but I have had little success in fixing it so far. Training with `keep_input=False` does work for PyTorch version 0.4.0, but appears to break training for PyTorch version >= 1.0.0. This, of course, corresponds to the new way of clearing and restoring the memory on the backward pass. I'll attempt to dig a little deeper tomorrow.

This implies that training with `keep_input=True` has to be done in PyTorch v0.4.0 for now.

**Comment by CorcovadoMing** [comment]
@silvandeleemput I've looked into original paper of RevNet in these days, and found maybe the current implementation is not able to achieve the space complexity O(1) due to we still need to store the output activator of each RevBlock into a buffer (https://github.com/silvandeleemput/memcnn/blob/master/memcnn/models/additive.py#L142)
(And I believe the NaN issue is due to the low-level storage is somehow sharing between `x` and `output`)

The optimized implementation should reconstruct the activators from the last RevBlock to beginning. However, PyTorch backward function is not able to pass the activators as argument (instead, it only propagates the gradient), so maybe that's why the current implementation still needs space complexity O(L)? unless we could leverage the activator passed from deeper layer in backward phase?

I'm not sure if it sounds reasonable to you? Feel free to correct me if I've misunderstood anything

**Comment by silvandeleemput** [comment]
@CorcovadoMing Thank you for giving it some thought. You have identified some of the pain points, but I will try to walk you through some of the details you might have missed:

The current implementation is able to achieve a memory complexity of O(1) for the activations during training (this is also shown by the tests for memory_savings.py). However, if you want to train it using the memory savings, MemCNN quickly gives NaN for the loss during training for PyTorch versions 1.0.0 and greater (This does not happen for PyTorch version 0.4.0). 

Restoring the activations in the current implementation is achieved at the moment by storing the input using `ctx.save_for_backward(x.data, output)` on forward and later restored by using x.set_() in the backward pass. Hence, the input and output tensor nodes are passed and available in the backward pass.

However, I am currently in the process of changing the way that this is handled, using a tensor hook and detaching the intermediate computation graph. And I think I have succeeded in implementing a simpler and more flexible way of wrapping invertible operations. I am currently going through testing and hope to be able to give you an update soon. This will probably be released in the next version of MemCNN and should fix the issue at hand.

**Comment by CorcovadoMing** [comment]
@silvandeleemput Thanks for your explanation, it's very helpful to understand this implementation :), through your last comment I believe the issue is caused by PyTorch fundamental storage management changes from 0.4 to 1.0+ 

I did some modifications for prove of concept and got a workaround working version:
- I release all inputs and outputs (by `x.storage().resize_(0)` excepts the following cases
- If previous layer is not RevBlock, the input should be kept
- if following layer is not RevBlock, the output should be kept
- Otherwise the output is reconstructed by propagated activator and nothing is being saved

Here comes an issue, how to propagate the activator via `Backward` PyTorch autograd function may be tricky. In my implementation, I use a global buffer. Obviously it's not a good idea and may not work in multi-gpu. but it did work in single-gpu!

Look forward to your next version. Thank you

**Comment by silvandeleemput** [comment]
@CorcovadoMing I have just merged my changes with master and I am in the process of releasing a new MemCNN version. This should resolve the training issues with memory savings enabled. I would like to encourage you to check out the readme for the updated features and example code.

**Changes**
* A complete refactor of MemCNN with changes to the API
  * Factored out the code responsible for the memory savings in a separate InvertibleModuleWrapper and reimplemented it using hooks
  * The InvertibleModuleWrapper allows for arbitrary invertible functions now (not just the additive and affine couplings)
  * The AdditiveBlock and AffineBlock have been refactored to AdditiveCoupling and AffineCoupling
  * The ReveribleBlock is now deprecated
  * The documentation and examples have been updated for the new API changes
